### PR TITLE
Improve creating tensor descriptor performance

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: cupy/cudnn.py [hunk 1]
```diff
@@ -99,13 +88,20 @@ def create_uninitialized_tensor_descriptor():
     return desc
 
 
+_nd_tensor_cache = {}
```
Comment (unnonouno) [inline cupy/cudnn.py]: Is this thread safe?

Comment (okuta) [inline cupy/cudnn.py]: Yes. Because this cache is only added new object.
And, the Python has GIL. Each primitive operation is thread safe.

Comment (gwtnb) [inline cupy/cudnn.py]: Is it safe to use a descriptor which was created by another thread?

Comment (okuta) [inline cupy/cudnn.py]: I do not know, but it seems to be read only

Comment (gwtnb) [inline cupy/cudnn.py]: Is it difficult to use thread-local caches?

Comment (okuta) [inline cupy/cudnn.py]: Yes, I can.

**Comment by unnonouno** [comment]
Do we need to use thread local variable?

**Comment by gwtnb** [comment]
Yes, we should use thread local variable at this time. We can make it non-thread-local later if the performance is unacceptable.

**Comment by unnonouno** [comment]
OK, i agree with you.

File: cupy/cudnn.py [hunk 2]
```diff
@@ -106,6 +103,12 @@ def create_tensor_nd_descriptor(arr):
         raise ValueError('cupy.cudnn supports c-contiguous arrays only')
     data_type = get_data_type(arr.dtype)
     shape = arr.shape
+    key = (data_type, shape)
+    getattr(_thread_local, 'default_backprop', True)
```
Comment (gwtnb) [inline cupy/cudnn.py]: What is the purpose of this line?

**Comment by unnonouno** [comment]
LGTM. @gwtnb do you have any concern?

**Comment by gwtnb** [comment]
No, it looks good.

**Pull request merged**